proof/refine/ARM/Syscall_R.thy

```diff
@@ -1156,98 +1179,121 @@ crunch valid_duplicates'[wp]: rescheduleRequired "\<lambda>s. vs_valid_duplicate
+                             elim!: st_tcb_weakenE)
+             apply (wp sts_st_tcb_at' set_thread_state_simple_sched_action
+                       set_thread_state_active_valid_sched set_thread_state_schact_is_rct_strong)
+            apply (rule_tac Q="\<lambda>rv. invs' and valid_invocation' rve'
+                                    and (\<lambda>s. thread = ksCurThread s)
+                                    and st_tcb_at' active' thread
+                                    and (\<lambda>s. ksSchedulerAction s = ResumeCurrentThread)
+                                    and (\<lambda>s. vs_valid_duplicates' (ksPSpace s))"
+                   in hoare_post_imp)
+             apply (clarsimp simp: ct_in_state'_def)
+             apply (frule (1) ct_not_ksQ)
+             apply (clarsimp)
+            apply (wp setThreadState_nonqueued_state_update
+                      setThreadState_st_tcb setThreadState_rct)
+           apply clarsimp
+           apply (wp lec_caps_to lsft_ex_cte_cap_to
+                  | simp add: split_def liftE_bindE[symmetric]
+                              ct_in_state'_def ball_conj_distrib
+                  | rule hoare_vcg_E_elim)+
    apply (clarsimp simp: msg_max_length_def word_bits_def)
+   apply (frule schact_is_rct_sane)
+   apply (frule invs_valid_objs)
+   apply (frule valid_objs_valid_tcbs)
+   apply (clarsimp simp: invs_def cur_tcb_def valid_state_def
+                         valid_sched_def valid_pspace_def ct_in_state_def simple_from_active)
    apply (erule st_tcb_ex_cap, clarsimp+)
    apply fastforce
-  apply (clarsimp)
-  apply (frule tcb_at_invs')
-  apply (clarsimp simp: invs'_def valid_state'_def
-                        ct_in_state'_def ct_not_inQ_def)
-  apply (frule(1) valid_queues_not_tcbQueued_not_ksQ)
-  apply (frule pred_tcb'_weakenE [where P=active' and P'=simple'], clarsimp)
-  apply (frule(1) st_tcb_ex_cap'', fastforce)
-  apply (clarsimp simp: valid_pspace'_def)
-  apply (frule(1) st_tcb_at_idle_thread')
-  apply (simp)
-  done *)
+  apply (clarsimp cong: conj_cong)
+  apply (subgoal_tac "is_schedulable_bool (cur_thread s) s")
+   apply (clarsimp simp: invs'_def valid_state'_def valid_pspace'_def
+                         cur_tcb'_def)
+   apply (frule valid_objs'_valid_tcbs')
+   apply (frule ct_active_cross, fastforce, fastforce, simp)
+   apply (clarsimp simp: ct_in_state'_def cong: conj_cong)
+   apply (frule pred_tcb'_weakenE [where P=active' and P'=simple'], clarsimp)
+   apply (frule(1) st_tcb_ex_cap'', fastforce)
+   apply (clarsimp simp: valid_pspace'_def schact_is_rct_def)
+   apply (frule state_relation_schact, simp)
+   apply (subgoal_tac "isSchedulable_bool (ksCurThread s') s'")
+    apply (clarsimp simp: isSchedulable_bool_def pred_map_conj)
+    apply (frule(1) st_tcb_at_idle_thread', simp)
+   apply (frule curthread_relation, simp)
+   apply (frule_tac t1="cur_thread s" in cross_relF[OF _ isSchedulable_bool_cross_rel];
```

Comment:
Maybe some of these subgoals here and above can be done with a `fastforce` instead? I've got no real preference, though